feat: add dedicated autocomplete profile type system (preview, dev only atm)

markijbema opened the pull request:
Add support for profile types to distinguish between chat and autocomplete provider configurations. This allows users to create a dedicated autocomplete profile that is automatically used for code completion while keeping it separate from chat profiles.

Key Features:
- Extensible profile type system: 'chat' (default) | 'autocomplete'
- Only ONE autocomplete profile allowed at a time (enforced by validation)
- Autocomplete profiles hidden from chat interface dropdown
- Status bar displays active autocomplete profile name
- Automatic migration sets existing profiles to 'chat' type

Backend Changes:
- Add ProfileType type and validation to provider-settings.ts
- Add validateAutocompleteConstraint() to ProviderSettingsManager
- Update GhostModel.reload() to prioritize autocomplete profiles
- Add profile name/type tracking methods to GhostModel
- Add migration logic for backward compatibility

UI Changes:
- Add profile type selector in providers settings (new & edit)
- Add '(Autocomplete)' badge to autocomplete profiles in list
- Filter autocomplete profiles from ChatTextArea profile selector
- Update GhostStatusBar to show 'Autocomplete: [Profile Name]'

Testing:
- Add comprehensive test coverage for all new functionality
- All existing tests pass (40 ProviderSettingsManager, 18 GhostModel, 165 ghost service, 36 ChatTextArea, 13 settings UI)

Breaking Changes: None
- Existing profiles automatically migrate to 'chat' type
- All changes are backward compatible

Closes #[issue-number]

## Context

<!-- Brief description of WHAT you’re doing and WHY. -->

## Implementation

<!--

Some description of HOW you achieved it. Perhaps give a high level description of the program flow. Did you need to refactor something? What tradeoffs did you take? Are there things in here which you’d particularly like people to pay close attention to?

Commit:
feat: add dedicated autocomplete profile type system

Add support for profile types to distinguish between chat and autocomplete
provider configurations. This allows users to create a dedicated autocomplete
profile that is automatically used for code completion while keeping it
separate from chat profiles.

Key Features:
- Extensible profile type system: 'chat' (default) | 'autocomplete'
- Only ONE autocomplete profile allowed at a time (enforced by validation)
- Autocomplete profiles hidden from chat interface dropdown
- Status bar displays active autocomplete profile name
- Automatic migration sets existing profiles to 'chat' type

Backend Changes:
- Add ProfileType type and validation to provider-settings.ts
- Add validateAutocompleteConstraint() to ProviderSettingsManager
- Update GhostModel.reload() to prioritize autocomplete profiles
- Add profile name/type tracking methods to GhostModel
- Add migration logic for backward compatibility

UI Changes:
- Add profile type selector in providers settings (new & edit)
- Add '(Autocomplete)' badge to autocomplete profiles in list
- Filter autocomplete profiles from ChatTextArea profile selector
- Update GhostStatusBar to show 'Autocomplete: [Profile Name]'

Testing:
- Add comprehensive test coverage for all new functionality
- All existing tests pass (40 ProviderSettingsManager, 18 GhostModel,
  165 ghost service, 36 ChatTextArea, 13 settings UI)

Breaking Changes: None
- Existing profiles automatically migrate to 'chat' type
- All changes are backward compatible

Closes #[issue-number]

markijbema commented on webview-ui/src/components/settings/__tests__/ApiConfigManager.spec.tsx:
```diff
@@ -42,6 +42,26 @@ vitest.mock("@/components/ui", () => ({
+			{children}
+		</select>
+	),
+	SelectTrigger: ({ children }: any) => <div className="select-trigger-mock">{children}</div>,
```
this just seems moved, cant we move it down again?

chrarnoldus commented on packages/types/src/provider-settings.ts:
```diff
@@ -37,6 +37,18 @@ import { toolUseStylesSchema } from "./kilocode/native-function-calling.js"
+export const profileTypeSchema = z.enum(profileTypes)
+export type ProfileType = z.infer<typeof profileTypeSchema>
+
+export const DEFAULT_PROFILE_TYPE: ProfileType = "chat"
```
ideally this is moved to a separate file

chrarnoldus commented on src/core/config/ProviderSettingsManager.ts:
```diff
@@ -304,6 +317,23 @@ export class ProviderSettingsManager {
+	/**
+	 * Migrate existing profiles to have profileType field defaulting to "chat"
+	 */
+	private async migrateProfileType(providerProfiles: ProviderProfiles) {
```
why do we have a migration? the code needs to be able to deal with missing profileTypes, as we can't realistically rely on new code setting it consistently.

markijbema replied:
removed

markijbema commented on packages/types/src/provider-settings.ts:
```diff
@@ -37,6 +37,18 @@ import { toolUseStylesSchema } from "./kilocode/native-function-calling.js"
+ * Profile Type System
+ */
+
+export const profileTypes = ["chat", "autocomplete"] as const
```
> or can you use autocomplete profiles there as well?

I havent prevented that (yet)

chrarnoldus reviewed (dismissed):
Approved pending release note

Commit:
add releasenote

EamonNerbonne committed:
Merge branch 'main' into mark/autocomplete-profile

Tricky merge conflict in GhostModel and less so ProviderSettingsManager

Commit:
manually unroll loop

Commit:
simplify

Commit:
invert if

Commit:
remove loop and simplify

there is always only one autocomplete profile if any

Commit:
fix tests

Commit:
no markers needed, this is our code

Commit:
remove isAutocompleteProfile

Commit:
remove useless methods

markijbema commented:
In testing i noticed that when you select the profile in the settings it will select the profile in the chat, which definitely isnt acceptable

Commit:
Merge branch 'main' into mark/autocomplete-profile

Commit:
feat: disable ModelSelector for autocomplete profiles

- Add isAutocomplete boolean to check profile type
- Disable dropdown when profile is autocomplete type
- Follow same pattern as isError condition
- Add comprehensive test coverage for autocomplete behavior

Commit:
Merge remote-tracking branch 'origin/main' into mark/autocomplete-profile

Commit:
make autocomplete model selection dev-only for now

Commit:
remove releasenote

EamonNerbonne committed:
Merge branch 'main' into mark/autocomplete-profile

Commit:
undo formatting changes

EamonNerbonne committed:
revert formatting changes

EamonNerbonne committed:
make these places more findable

EamonNerbonne committed:
Merge pull request #3858 from Kilo-Org/eamon/idea

make these places more findable

EamonNerbonne committed:
Extract commonality

EamonNerbonne committed:
Merge remote-tracking branch 'origin/main' into mark/autocomplete-profile

EamonNerbonne committed:
Merge pull request #3863 from Kilo-Org/eamon/commonality

Extract commonality

markijbema commented on src/services/ghost/GhostModel.ts:
```diff
@@ -27,9 +31,25 @@ export class GhostModel {
+			}
+
+			this.loaded = true
+			return true
```
we don't do this for autodetection so if someone has both kilo for free grok-code and openrouter with credits they can use openrouter

EamonNerbonne committed:
Reduce diff size

EamonNerbonne committed:
rename to reduce diff siez

EamonNerbonne committed:
Revert to main; reapply core change to reduce diff size

EamonNerbonne committed:
kilocode changes markers

EamonNerbonne commented on webview-ui/src/components/settings/__tests__/ApiConfigManager.spec.tsx:
```diff
@@ -205,8 +205,9 @@ describe("ApiConfigManager", () => {
 
-		// Find input and enter new name
-		const input = screen.getByDisplayValue("Default Config")
+		// Find input in rename form specifically
```
resolved.

EamonNerbonne committed:
Merge branch 'main' into mark/autocomplete-profile

EamonNerbonne reviewed (approved):
So, I don't see urgent issues anymore.

However, the CI does, the webview tests are failing, and have been doing so for quite some time.

Commit:
fix: resolve test timeout by fixing react-query mock pattern

The test was using an inline factory mock for @tanstack/react-query which
doesn't properly auto-mock all exports. This caused issues with the
usePromptHistory hook which uses useTaskHistory, which requires
taskHistoryVersion from useExtensionState.

Changes:
- Use vi.mock('@tanstack/react-query') for auto-mocking
- Configure useQuery mock in beforeEach
- Add taskHistoryVersion: 0 and clineMessages: [] to mock return values
- Minor refactor in ChatTextArea.tsx for explicit null check

Commit:
Merge remote-tracking branch 'origin/main' into mark/autocomplete-profile

Commit:
fix: handle missing apiProvider in autocomplete profile

- Fixed type error in GhostModel.ts where useProfile was called with 2
  arguments instead of 3 for autocomplete profiles
- Added check to ensure apiProvider is defined before using the profile
- Fixed tests that incorrectly used Object.keys() on a Map instead of
  spreading the Map keys

EamonNerbonne commented on src/services/ghost/__tests__/GhostModel.spec.ts:
```diff
@@ -348,7 +348,7 @@ describe("GhostModel", () => {
 
 			it("stores and returns profile name after loading", async () => {
-				const supportedProviders = Object.keys(AUTOCOMPLETE_PROVIDER_MODELS)
+				const supportedProviders = [...AUTOCOMPLETE_PROVIDER_MODELS.keys()]
```
noise?  or signal?

(fine either way, just curious)

markijbema replied:
Opus said

[src/services/ghost/__tests__/GhostModel.spec.ts](vscode-webview://1n3era6ha6ld3ktpthg0f26upj33ab2kh1oducguj98qv9gqc1t8/src/services/ghost/__tests__/GhostModel.spec.ts:351-410) - Fixed three tests that incorrectly used Object.keys(AUTOCOMPLETE_PROVIDER_MODELS) on a Map (which returns an empty array). Changed to [...AUTOCOMPLETE_PROVIDER_MODELS.keys()] to properly get the provider keys.
